File: slack-reader/src/test/java/dev/number6/slackreader/dagger/FakeAmazonSnsModule.java
package dev.number6.slackreader.dagger;

import com.amazonaws.services.sns.AbstractAmazonSNS;
import com.amazonaws.services.sns.AmazonSNS;
import com.amazonaws.services.sns.model.PublishRequest;
import com.amazonaws.services.sns.model.PublishResult;
import dagger.Module;
import dagger.Provides;

import javax.inject.Singleton;
import java.util.ArrayList;
import java.util.Collection;

@Module
public class FakeAmazonSnsModule {

    @Provides
    public AmazonSNS providesAmazonSNS(FakeAmazonSns sns) {

        return sns;
    }

    @Provides
    @Singleton
    FakeAmazonSns providesFakeAmazonSns(){
        return new FakeAmazonSns();
    }

    public class FakeAmazonSns extends AbstractAmazonSNS {

        final Collection<String> published = new ArrayList<>();

        public Collection<String> getPublishedMessages() {
            return published;
        }

        @Override
        public PublishResult publish(PublishRequest request) {
            published.add(request.getMessage());
            return new PublishResult();
        }
    }
}

File: slack-reader/src/test/kotlin/dev/number6/slackreader/dagger/FakeAmazonSnsModule.kt
package dev.number6.slackreader.dagger

import com.amazonaws.services.sns.AbstractAmazonSNS
import com.amazonaws.services.sns.AmazonSNS
import com.amazonaws.services.sns.model.PublishRequest
import com.amazonaws.services.sns.model.PublishResult
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class FakeAmazonSnsModule {
    @Provides
    fun providesAmazonSNS(sns: FakeAmazonSns): AmazonSNS {
        return sns
    }

    @Provides
    @Singleton
    fun providesFakeAmazonSns(): FakeAmazonSns {
        return FakeAmazonSns()
    }

    inner class FakeAmazonSns : AbstractAmazonSNS() {
        val published: MutableCollection<String> = mutableListOf()
        fun getPublishedMessages(): MutableCollection<String> {
            return published
        }

        override fun publish(request: PublishRequest): PublishResult {
            published.add(request.message)
            return PublishResult()
        }
    }
}